File: PlayerPRO Player/PPMusicListDragClass.m
//
//  PPMusicListDragClass.m
//  PPMacho
//
//  Created by C.W. Betts on 12/8/13.
//
//

#import "PPMusicListDragClass.h"
#import "UserDefaultKeys.h"

@interface PPMusicListDragClass ()
@property (readwrite, copy) NSIndexSet *theIndexSet;
@end

@implementation PPMusicListDragClass

- (instancetype)initWithIndexSet:(NSIndexSet*)theSet
{
	if (self = [super init]) {
		self.theIndexSet = theSet;
	}
	return self;
}

+ (instancetype)dragWithIndexSet:(NSIndexSet*)theSet
{
	return [[[self class] alloc] initWithIndexSet:theSet];
}

+ (NSPasteboardReadingOptions)readingOptionsForType:(NSString *)type pasteboard:(NSPasteboard *)pasteboard
{
	if ([type isEqualToString:PPMLDCUTI])
		return NSPasteboardReadingAsKeyedArchive;
	else
		return NSPasteboardReadingAsData;
}

static NSArray *UTIArray;
static dispatch_once_t dragOnceToken;
static const dispatch_block_t initUTIArray = ^{
	UTIArray = @[PPMLDCUTI];
};

+ (NSArray *)readableTypesForPasteboard:(NSPasteboard *)pasteboard
{
	dispatch_once(&dragOnceToken, initUTIArray);
	return UTIArray;
}

- (NSArray *)writableTypesForPasteboard:(NSPasteboard *)pasteboard
{
	dispatch_once(&dragOnceToken, initUTIArray);
	return UTIArray;
}

- (id)pasteboardPropertyListForType:(NSString *)type
{
	if ([type isEqualToString:PPMLDCUTI])
		return [NSKeyedArchiver archivedDataWithRootObject:self];
	else
		return nil;
}

- (void)encodeWithCoder:(NSCoder *)aCoder
{
	[aCoder encodeObject:_theIndexSet forKey:PPMLDCUTI];
}

- (id)initWithCoder:(NSCoder *)aDecoder
{
	return self = [self initWithIndexSet:[aDecoder decodeObjectForKey:PPMLDCUTI]];
}

@end

File: PlayerPRO Player/PPMusicListDragClass.swift
//
//  PPMusicListDragClass.swift
//  PPMacho
//
//  Created by C.W. Betts on 7/24/14.
//
//

import Cocoa

private let UTIArray = [PPMLDCUTI];

class PPMusicListDragClass: NSObject, NSPasteboardReading, NSPasteboardWriting, NSSecureCoding {
	private(set) var theIndexSet: NSIndexSet
	init(indexSet: NSIndexSet!) {
		theIndexSet = indexSet
		
		super.init()
	}
	
	class func readableTypesForPasteboard(pasteboard: NSPasteboard!) -> [AnyObject]! {
		return UTIArray
	}
	
	func writableTypesForPasteboard(pasteboard: NSPasteboard!) -> [AnyObject]! {
		return UTIArray
	}
	
	class func readingOptionsForType(type: String!, pasteboard: NSPasteboard!) -> NSPasteboardReadingOptions {
		if (type == PPMLDCUTI) {
			return .AsKeyedArchive;
		} else {
			return .AsData;
		}
	}
	
	init() {
		theIndexSet = NSIndexSet()
		
		super.init()
	}
	
	convenience init(pasteboardPropertyList propertyList: AnyObject!, ofType type: String!) {
		if (type == PPMLDCUTI) {
			var unArchive = NSUnarchiver(forReadingWithData: propertyList as NSData)
			self.init(indexSet:unArchive.decodeObjectForKey(PPMLDCUTI) as NSIndexSet)
		} else {
			self.init()
		}
	}
	
	func pasteboardPropertyListForType(type: String) -> AnyObject? {
		if (type == PPMLDCUTI) {
			return NSKeyedArchiver.archivedDataWithRootObject(self);
		} else {
			return nil;
		}
	}

	class func supportsSecureCoding() -> Bool {
		return true
	}
	
	func encodeWithCoder(aCoder: NSCoder) {
		aCoder.encodeObject(theIndexSet, forKey: PPMLDCUTI)
	}
	
	convenience init(coder aDecoder: NSCoder!) {
		self.init(indexSet:aDecoder.decodeObjectForKey(PPMLDCUTI) as NSIndexSet)
	}
	

}
